go-controller/pkg/util/dpu_annotations.go

```diff
@@ -56,75 +59,124 @@ type DPUConnectionStatus struct {
 	Reason string `json:"Reason,omitempty"`
 }
 
-func (scd *DPUConnectionDetails) FromPodAnnotation(podAnnot map[string]string) error {
-	if annot, ok := podAnnot[DPUConnectionDetailsAnnot]; ok {
-		if err := json.Unmarshal([]byte(annot), scd); err != nil {
-			return fmt.Errorf("failed to unmarshal DPUConnectionDetails. %v", err)
+// UnmarshalPodDPUConnDetailsAllNetworks returns the DPUConnectionDetails map of all networks from the given Pod annotation
+func UnmarshalPodDPUConnDetailsAllNetworks(annotations map[string]string) (map[string]DPUConnectionDetails, error) {
+	podDcds := make(map[string]DPUConnectionDetails)
+	ovnAnnotation, ok := annotations[DPUConnectionDetailsAnnot]
+	if ok {
+		if err := json.Unmarshal([]byte(ovnAnnotation), &podDcds); err != nil {
+			// DPU connection details annotation could be in the legacy format
+			var legacyScd DPUConnectionDetails
+			if err := json.Unmarshal([]byte(ovnAnnotation), &legacyScd); err == nil {
+				podDcds[types.DefaultNetworkName] = legacyScd
+			} else {
+				return nil, fmt.Errorf("failed to unmarshal OVN pod %s annotation %q: %v",
+					DPUConnectionDetailsAnnot, annotations, err)
+			}
 		}
-		return nil
 	}
-	return fmt.Errorf("failed to get DPUConnectionDetails, pod annotation \"%s\" does not exist",
-		DPUConnectionDetailsAnnot)
+	return podDcds, nil
 }
 
-func (scd *DPUConnectionDetails) SetPodAnnotation(podAnnotator kube.Annotator) error {
-	data, err := json.Marshal(scd)
+// MarshalPodDPUConnDetails adds the pod's connection details of the specified network to the corresponding pod annotation.
+func MarshalPodDPUConnDetails(pannotations *map[string]string, dcd *DPUConnectionDetails, netName string) error {
+	var err error
+	annotations := *pannotations
+	// in case the given pod annotation is nil, set the updated pod annotation back
+	if annotations == nil {
+		annotations = make(map[string]string)
+		*pannotations = annotations
+	}
```

Comment:
I think that probably a more canonical & explicit way of doing this is
```
func MarshalPodDPUConnDetails(pannotations map[string]string, dcd *DPUConnectionDetails, netName string) (map[string]string, error)
```
so instead of passing a pointer, just returning the annotations back.

This comment applies to all the places where a similar thing is being done.

Reply:
I agree; I'd rather see the annotations returned than mutate the input.